apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.PostPolicy.cs

```diff
@@ -0,0 +1,199 @@
+            /// The returned instance will contain a range condition for <paramref name="element"/>.
+            /// </summary>
+            /// <typeparam name="TElement">The type of the element this condition will apply to.
+            /// It needs to implement <see cref="ISupportsRange{TValue}"/>.</typeparam>
+            /// <typeparam name="TValue">The type of the values defining the range.</typeparam>
+            /// <param name="element">The element for which to set a range condition.</param>
+            /// <param name="min">The range lower bound.</param>
+            /// <param name="max">The range upper bound.</param>
+            /// <returns>A new instance of <see cref="PostPolicy"/> which contains the same conditions
+            /// as this one, except for the condition applied to <paramref name="element"/>.</returns>
+            public PostPolicy WithRange<TElement, TValue>(TElement element, TValue min, TValue max)
+                where TElement : class, ISupportsRange<TValue>
+            {
+                PostPolicy postPolicy = new PostPolicy(_expiration);
+                foreach (var entry in _conditions)
+                {
+                    postPolicy.AddOrReplace(entry.Key, entry.Value);
+                }
+                postPolicy.AddOrReplace(element, new InRange<TElement, TValue>(element, min, max));
+                return postPolicy;
+            }
+
+            /// <summary>
+            /// Returns a new <see cref="PostPolicy"/> which is the same as this one
+            /// except for its expiration.
+            /// </summary>
+            /// <param name="expiration">The new expiration for the policy.</param>
+            /// <returns>A new instance of <see cref="PostPolicy"/> which isthe same as this one
+            /// except for its expiration.</returns>
+            public PostPolicy WithExpiration(DateTime expiration)
+            {
+                PostPolicy postPolicy = new PostPolicy(expiration);
+                foreach (var entry in _conditions)
+                {
+                    postPolicy.AddOrReplace(entry.Key, entry.Value);
+                }
+                return postPolicy;
+            }
+
+            internal string ToJsonString()
+            {
+                GaxPreconditions.CheckState(_expiration.HasValue, "Can't use a policy with no expiration.");
+
+                string expirationText = _expiration.Value.ToUniversalTime().ToString("YYYY-MM-DD'T'HH:MM:SS'Z'", CultureInfo.InvariantCulture);
+                string conditions = string.Join(",", _conditions.Values.Select(cond => cond.ToJsonString()));
+
+                // {\"conditions\":[{\"key\":\"test-object\"},{\"x-goog-date\":\"20200123T043530Z\"},{\"x-goog-credential\":\"test-iam-credentials@dummy-project-id.iam.gserviceaccount.com/20200123/auto/storage/goog4_request\"},{\"x-goog-algorithm\":\"GOOG4-RSA-SHA256\"}],\"expiration\":\"2020-01-23T04:35:40Z\"}
```

Comment:
conditions willl now include `{\"bucket\":\"bucket-name\"}` to protect against attack vector discussed.

Reply:
Yes, my implementation already includes them, that's why my tests were failing. Will update tests now and on next push they should be green.